common/src/main/java/org/apache/falcon/entity/DatasourceHelper.java

```diff
@@ -148,16 +152,168 @@ public static String buildJceksProviderPath(URI credURI) {
+            }
+        }
+        return false;
+    }
+
+    public static Credential getCredential(Datasource db) {
+        return getCredential(db, null);
+    }
+
+    public static Credential getCredential(Datasource db, Interfacetype interfaceType) {
+        if (interfaceType == null) {
+            return db.getInterfaces().getCredential();
+        } else {
+            for(Interface iface : db.getInterfaces().getInterfaces()) {
+                if (iface.getType() == interfaceType) {
+                    return iface.getCredential();
+                }
+            }
+        }
+        return null;
+    }
+
+    public static void validateCredential(Credential cred) throws FalconException {
+        if (cred == null) {
+            return;
+        }
+        switch (cred.getType()) {
+        case PASSWORD_TEXT:
+            if (StringUtils.isBlank(cred.getUserName()) || StringUtils.isBlank(cred.getPasswordText())) {
+                throw new FalconException(String.format("Credential type '%s' missing tags '%s' or '%s'",
+                    cred.getType().value(), "userName", "passwordText"));
+            }
+            break;
+        case PASSWORD_FILE:
+            if (StringUtils.isBlank(cred.getUserName()) || StringUtils.isBlank(cred.getPasswordFile())) {
+                throw new FalconException(String.format("Credential type '%s' missing tags '%s' or '%s'",
+                    cred.getType().value(), "userName", "passwordFile"));
+            }
+            break;
+        case PASSWORD_ALIAS:
+            if (StringUtils.isBlank(cred.getUserName()) || (cred.getPasswordAlias() == null)
+                || StringUtils.isBlank(cred.getPasswordAlias().getAlias())
+                || StringUtils.isBlank(cred.getPasswordAlias().getProviderPath())) {
+                throw new FalconException(String.format("Credential type '%s' missing tags '%s' or '%s' or %s'",
+                    cred.getType().value(), "userName", "alias", "providerPath"));
+            }
+            break;
+        default:
+            throw new FalconException(String.format("Unknown Credential type '%s'", cred.getType().value()));
+        }
+    }
+
+    private static boolean isSameString(String ep1, String ep2) {
+        if ((StringUtils.isBlank(ep1)) && (StringUtils.isBlank(ep2))) {
+            return true;
+        }
+        if ((StringUtils.isBlank(ep1)) && (StringUtils.isBlank(ep2))) {
+            return false;
+        }
+        return ep1.equals(ep2);
+
+    }
+    /**
      * Return the Interface endpoint for the interface type specified in the argument.
      *
      * @param db
      * @param type - can be read-only or write
      * @return
      */
-    private static String getInterface(Datasource db, Interfacetype type) {
+    private static String getInterfaceEndpoint(Datasource db, Interfacetype type) {
+        if (getInterface(db, type) != null) {
+            return getInterface(db, type).getEndpoint();
```

Comment:
Better to record the interface getInterface(db, type) so it doesn't need to be computed twice.